File: 921.Minimum Add to Make Parentheses Valid/921_Minimum Add to Make Parentheses Valid.swift
//
//  test.swift
//  LeetcodeWithSwift
//
//  Created by Chris Cheung on 2018/11/14.
//  Copyright © 2018 OBJCHRIS. All rights reserved.
//

import Foundation

class Solution {
    func minAddToMakeValid(_ S: String) -> Int {
        var ans = 0;
        var arr = [Character]()
        for c in S {
            if c == '(' {
                arr.append(c)
            } else if c == ')' && !arr.isEmpty {
                arr.popLast()
            } else {
                ans += 1
            }
        }
        ans += arr.count
        return ans
    }
}

File: 其他/921.Minimum Add to Make Parentheses Valid/921_Minimum Add to Make Parentheses Valid.cpp
//
//  main.cpp
//  LeetcodeWithCPP
//
//  Created by Chris Cheung on 2018/11/13.
//  Copyright © 2018 OBJCHRIS. All rights reserved.
//

#include <stdio.h>
#include <iostream>
#include <string>
#include <stack>

using namespace std;

class Solution {
public:
    int minAddToMakeValid(string S) {
        stack<char> ans;
        for (int i=0; i<S.size(); i++) {
            if(ans.empty()) {
                ans.push(S[i]);
                continue;
            }
            if(S[i] == ')' && ans.top() == '(') {
                ans.pop();
                continue;
            }
            ans.push(S[i]);
        }
        return ans.size();
    }
    
private:
    int minAddToMakeValid_fast(string S) {
        stack<int> ss;
        int res = 0;
        for(int i=0;i<S.size();i++){
            if(S[i] == '('){
                ss.push(S[i]);
            }else if(S[i] == ')' && !ss.empty()){
                ss.pop();
            }else{
                res++;
            }
        }
        res = res + ss.size();
        return res;
    }
};


int main() {
    Solution s = Solution();
    int i = s.minAddToMakeValid("()))((");
    cout << i;
}
